harness/determined/launch/deepspeed.py

```diff
@@ -7,34 +7,38 @@
 import logging
 import os
 import pathlib
+import shlex
 import subprocess
 import sys
 import time
-from typing import List
+from typing import List, Optional
 
+import deepspeed
 from deepspeed.launcher.runner import DEEPSPEED_ENVIRONMENT_NAME
+from packaging import version
 
 import determined as det
 from determined import constants, util
 from determined.common import api
 from determined.common.api import certs
 
-hostfile_path = "/tmp/hostfile.txt"
+deepspeed_version = version.parse(deepspeed.__version__)
 
 
 def create_hostlist_file(
-    hostfile_path: pathlib.Path, num_proc_per_machine: int, ip_addresses: List[str]
+    hostfile_path: Optional[str], num_proc_per_machine: int, ip_addresses: List[str]
 ) -> str:
     trial_runner_hosts = ip_addresses.copy()
     # In the single node case, deepspeed doesn't use pdsh so we don't need to launch sshd.
     # Instead, the deepspeed launcher will use localhost as the chief worker ip.
     if len(ip_addresses) == 1:
         trial_runner_hosts[0] = "localhost"
 
-    os.makedirs(hostfile_path.parent, exist_ok=True)
-    with open(hostfile_path, "w") as hostfile:
-        lines = [f"{host} slots={num_proc_per_machine}\n" for host in trial_runner_hosts]
-        hostfile.writelines(lines)
+    if hostfile_path is not None:
+        os.makedirs(pathlib.Path(hostfile_path).parent, exist_ok=True)
```

Comment:
nit: stick to either str or pathlib, don't go back and forth.

Replace `pathlib.Path(hostfile_path).parent` with `os.path.dirname(hostfile_path)`